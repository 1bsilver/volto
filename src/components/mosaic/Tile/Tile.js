/**
 * Tile component.
 * @module components/Tile
 */

import React, { PropTypes } from 'react';

import { Editor } from 'components';

import styles from './Tile.scss';

/**
 * Tile component class.
 * @function Tile
 * @param {Object} props Component properties.
 * @param {Object} props.content Content of the tile.
 * @param {number} props.width Width of the tile.
 * @param {bool} props.selected True if tile is selected.
 * @param {number} props.row Row index.
 * @param {number} props.column Column index.
 * @param {func} props.selectTile Select tile method.
 * @param {func} props.setTileContent Set tile content method.
 * @returns {string} Markup of the tile.
 */
const Tile = ({ content, width, row, column, selected, selectTile, setTileContent }) =>
  <div className={`${styles.tile} col-xs-${width} ${selected ? styles.selected : ''}`}>
    <div onClick={() => selectTile(row, column)} className={styles.content}>
      <Editor content={content} setContent={newContent => setTileContent(row, column, newContent)} />
    </div>
  </div>;

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
Tile.propTypes = {
  content: PropTypes.object.isRequired,
  width: PropTypes.number.isRequired,
  row: PropTypes.number.isRequired,
  column: PropTypes.number.isRequired,
  selected: PropTypes.bool.isRequired,
  selectTile: PropTypes.func.isRequired,
  setTileContent: PropTypes.func.isRequired,
};

export default Tile;
