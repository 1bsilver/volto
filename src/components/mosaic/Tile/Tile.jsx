/**
 * Tile component.
 * @module components/Tile
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Label } from 'semantic-ui-react';
import { DragSource, DropTarget } from 'react-dnd';
import { findDOMNode } from 'react-dom';

import Editor from 'draft-js-editor';

/* eslint jsx-a11y/no-static-element-interactions: 0 */

/**
 * Tile component class.
 * @function Tile
 * @param {Object} props Component properties.
 * @param {Object} props.content Content of the tile.
 * @param {string} props.type Type of the tile.
 * @param {bool} props.selected True if tile is selected.
 * @param {string} props.hovered Hovered direction.
 * @param {number} props.row Row index.
 * @param {number} props.column Column index.
 * @param {number} props.tile Tile index.
 * @param {func} props.selectTile Select tile method.
 * @param {func} props.setHovered Set hovered tile method.
 * @param {func} props.setTileContent Set tile content method.
 * @returns {string} Markup of the tile.
 */
const Tile = ({
  content,
  type,
  selected,
  hovered,
  row,
  column,
  tile,
  selectTile,
  setTileContent,
  connectDragSource,
  connectDropTarget,
  isDragging,
}) =>
  connectDropTarget(
    connectDragSource(
      <div
        className={`tile ${selected ? 'selected' : ''} ${isDragging ? 'dragging' : ''} ${hovered || ''}`}
        onClick={() => selectTile(row, column, tile)}
      >
        <Label color="blue" pointing="below">{type}</Label>
        {!__SERVER__ &&
          <Editor
            onChange={newContent =>
              setTileContent(row, column, tile, newContent)}
            editorState={content}
          />}
      </div>,
    ),
  );

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
Tile.propTypes = {
  content: PropTypes.object.isRequired,
  type: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  hovered: PropTypes.string,
  row: PropTypes.number.isRequired,
  column: PropTypes.number.isRequired,
  tile: PropTypes.number.isRequired,
  selectTile: PropTypes.func.isRequired,
  setHovered: PropTypes.func.isRequired,
  setTileContent: PropTypes.func.isRequired,
};

/**
 * Default props.
 * @property {Object} defaultProps Default properties.
 * @static
 */
Tile.defaultProps = {
  hovered: null,
};

export default DropTarget(
  'tile',
  {
    hover(props, monitor, component) {
      const dragRow = monitor.getItem().row;
      const dragColumn = monitor.getItem().column;
      const hoverRow = props.row;
      const hoverColumn = props.column;

      if (dragRow === hoverRow && dragColumn === hoverColumn) {
        return;
      }

      const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();
      const clientOffset = monitor.getClientOffset();

      const height = hoverBoundingRect.top - hoverBoundingRect.bottom;
      const y = clientOffset.y - hoverBoundingRect.bottom - height / 2;
      const yFactor = y / (height / 2);

      const width = hoverBoundingRect.right - hoverBoundingRect.left;
      const x = clientOffset.x - hoverBoundingRect.left - width / 2;
      const xFactor = x / (width / 2);

      let direction;
      if (Math.abs(xFactor) > Math.abs(yFactor)) {
        direction = xFactor > 0 ? 'right' : 'left';
      } else {
        direction = yFactor > 0 ? 'top' : 'bottom';
      }
      if (props.hovered !== direction) {
        props.setHovered(props.row, props.column, props.tile, direction);
      }
    },
    drop(props) {
      props.setHovered(-1, -1, -1, null);
    },
  },
  connect => ({
    connectDropTarget: connect.dropTarget(),
  }),
)(
  DragSource(
    'tile',
    {
      beginDrag(props) {
        props.selectTile(-1, -1);
        return {
          row: props.row,
          column: props.column,
          tile: props.tile,
        };
      },
    },
    (connect, monitor) => ({
      connectDragSource: connect.dragSource(),
      connectDragPreview: connect.dragPreview(),
      isDragging: monitor.isDragging(),
    }),
  )(Tile),
);
