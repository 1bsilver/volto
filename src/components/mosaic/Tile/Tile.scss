@import "../../../theme/variables.scss";

.tile {
  border: solid 2px transparent;
  cursor: move;
  padding: 0;
  &:hover {
    border: dashed 2px $current-tile-border-color;
  }
}

.content {
  z-index: 100;
}

.selected {
  border: solid 2px $current-tile-border-color;
  cursor: default;
  background-color: $edit-color;
  &:hover {
    border: solid 2px $current-tile-border-color;
  }
}
