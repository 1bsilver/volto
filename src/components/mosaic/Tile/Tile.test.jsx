import React from 'react';
import renderer from 'react-test-renderer';
import Tile from './Tile';

jest.mock('draft-js-editor',
  () => jest.fn(() => <div />));
global.__SERVER__ = true; // eslint no-underscore-dangle: 0

test('renders a tile component', () => {
  const component = renderer.create(
    <Tile
      content={ { data: '<h1>Hello World!</h1>'} }
      width={4}
      row={0}
      column={0}
      selected={true}
      selectTile={() => {}}
      setTileContent={() => {}}
    />
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
