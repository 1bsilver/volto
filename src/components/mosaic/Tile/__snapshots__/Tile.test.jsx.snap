// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a tile component 1`] = `
<div>
  <div
    className="tile selected  "
    onClick={[Function]}
  >
    <div
      className="ui blue pointing below label"
      onClick={[Function]}
    >
      Title
    </div>
  </div>
</div>
`;
